stacker/dag/__init__.py

```diff
@@ -0,0 +1,264 @@
+import logging
+from copy import copy, deepcopy
+from collections import deque
+
+logger = logging.getLogger(__name__)
+
+try:
+    from collections import OrderedDict
+except:
+    from ordereddict import OrderedDict
+
+
+class DAGValidationError(Exception):
+    pass
+
+
+class DAG(object):
+    """ Directed acyclic graph implementation. """
+
+    def __init__(self):
+        """ Construct a new DAG with no nodes or edges. """
+        self.reset_graph()
+
+    def add_node(self, node_name, graph=None):
+        """ Add a node if it does not exist yet, or error out. """
+        if not graph:
```

Comment:
Do you have any idea why this library does this so often?  I haven't seen this pattern used really ever before.  I thought maybe for testing, but it seems like a messy way to handle that.

Reply:
No clue. It seems unnecessary. I'll try to add a commit that removes this.

Reply:
Thanks, yeah, I couldn't figure it out - I thought it'd be a test thing, but just now looking at tests I don't think it is.  Strange.